### PR TITLE
[Tests] Make it possible and easy to test exception messages

### PR DESCRIPTION
**Pull request by segiddins**
Yay macros!

\c @tgoyne @bdash

### CODE REVIEW

File: Realm/Tests/RLMAssertions.h [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+//
+////////////////////////////////////////////////////////////////////////////
+
+#import <XCTest/XCTest.h>
+
+#define RLMAssertThrows(expression, ...) \
+RLMPrimativeAssertThrows(self, expression,  __VA_ARGS__)
+
+#define RLMPrimativeAssertThrows(self, expression, format...) \
+({ \
```
Comment (tgoyne) [inline Realm/Tests/RLMAssertions.h]: Please indent the body of macros normally.

File: Realm/Tests/RLMAssertions.h [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+__caughtException; \
+})
+
+#define RLMAssertMatches(expression, regex, ...) \
+RLMPrimativeAssertMatches(self, expression, regex,  __VA_ARGS__)
+
+#define RLMPrimativeAssertMatches(self, expression, regex, format...) \
+({ \
+NSString *string = (expression);\
+NSRegularExpression *__regex = [NSRegularExpression regularExpressionWithPattern: regex options: 0 error:nil];\
```
Comment (tgoyne) [inline Realm/Tests/RLMAssertions.h]: Names beginning with a double underscore are reserved for the implementation.

Comment (segiddins) [inline Realm/Tests/RLMAssertions.h]: what do you mean, `reserved for the implementation`? This is the implementation, no?

Comment (tgoyne) [inline Realm/Tests/RLMAssertions.h]: The compiler and standard library.

Comment (segiddins) [inline Realm/Tests/RLMAssertions.h]: I was just copying `XCTest`'s style, which uses `__` for these internal variables. I can change them

File: Realm/Tests/RLMAssertions.h [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+__caughtException = exception; \
+}\
+if (!__caughtException) { \
+_XCTRegisterFailure(self, _XCTFailureDescription(_XCTAssertion_Throws, 0, @#expression), format); \
+} \
+__caughtException; \
+})
+
+#define RLMAssertMatches(expression, regex, ...) \
+RLMPrimativeAssertMatches(self, expression, regex,  __VA_ARGS__)
```
Comment (bdash) [inline Realm/Tests/RLMAssertions.h]: s/Primative/Primitive/

**Comment by tgoyne** [comment]
:+1:

File: Realm/Tests/RLMAssertions.h [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+#define RLMAssertMatches(expression, regex, ...) \
+    RLMPrimitiveAssertMatches(self, expression, regex,  __VA_ARGS__)
+
+#define RLMPrimitiveAssertMatches(self, expression, regexString, format...) \
+({ \
+    NSString *string = (expression); \
+    NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern: regexString options: 0 error:nil]; \
+    if ([regex numberOfMatchesInString:string options:0 range:NSMakeRange(0, string.length)] == 0) { \
+        _XCTRegisterFailure(self, [_XCTFailureDescription(_XCTAssertion_True, 0, @#expression @" (EXPR_STRING) matches " @#regexString) stringByReplacingOccurrencesOfString:@"EXPR_STRING" withString:string ?: @"<nil>"], format); \
```
Comment (jpsim) [inline Realm/Tests/RLMAssertions.h]: shouldn't this be "does not match" instead of "matches" since it's a failure?

Comment (segiddins) [inline Realm/Tests/RLMAssertions.h]: `/Users/segiddins/Development/Realm/realm-cocoa/Realm/Tests/DynamicTests.m:113: error: -[DynamicTests testDynamicProperties] : ((exception.reason (Invalid property name) matches @"name\\.") is true) failed`

**Comment by jpsim** [comment]
:+1:

**Pull request merged**